handlers/proxy_handler.cc

```diff
@@ -0,0 +1,240 @@
+
+	modified_request->SetURI(modified_uri);
+	modified_request->SetHeader("Connection", "close");
+	modified_request->SetHeader("Host", m_host_path_ + ":" + m_port_path_);
+}
+
+// given a location string like "http://www.ucla.edu/some/path", extract "www.ucla.edu" and "/some/path"
+void ProxyHandler::ParseRedirectLocation(std::string location, std::string* new_path, std::string* new_host) {
+	location = boost::algorithm::trim_copy(location);
+
+	// strip "http://", if present
+	const size_t http_size = 7; // "http://"
+	if (location.size() > http_size && location.substr(0, http_size) == "http://") {
+		location = location.substr(http_size);
+	}
+
+	const size_t https_size = 8; // "https://"
+	if (location.size() > https_size && location.substr(0, https_size) == "https://") {
+		printf("ProxyHandler: warning: redirect specified https, which is unsupported.\n");
+		location = location.substr(https_size);
+	}
+
+	// Location will be in form `newhost/new/path`, we want to split into 'newhost' + '/new/path'
+	auto pathPos = location.find('/');
+	if (pathPos != std::string::npos) {
+		*new_host = location.substr(0, pathPos);
+		*new_path = location.substr(pathPos);
+	} else {
+		*new_host = location;
+		*new_path = "/"; // no path given
+	}
+}
+
+void ProxyHandler::ConnectSocketToEndpoint(tcp::socket* socket, std::string host, std::string port) {
+	tcp::resolver resolver(io_service_);
+	tcp::resolver::query query(host, port);
+	tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
+	boost::asio::connect(*socket, endpoint_iterator);
+}
+
+void ProxyHandler::WriteToSocket(tcp::socket* socket, const std::string& requestStr) {
+	boost::asio::streambuf reqBuf;
+	std::ostream request_stream(&reqBuf);
+	request_stream << requestStr;
+	boost::asio::write(*socket, reqBuf);
+}
+
+boost::system::error_code ProxyHandler::SocketReadToEOF(tcp::socket* socket, boost::asio::streambuf* buf, std::string* data) {
+	boost::system::error_code error;
+	while (true) {
+		boost::asio::read(*socket, *buf, boost::asio::transfer_at_least(1), error);
+		if (buf->size() == 0) {
+			break;
+		}
+		auto bufs = buf->data();
+		std::string bodyChunk(boost::asio::buffers_begin(bufs), boost::asio::buffers_begin(bufs) + buf->size());
+		*data += bodyChunk;
+		buf->consume(buf->size());
+	}
+	return error;
+}
+
+boost::system::error_code ProxyHandler::SocketReadUntil(
+		tcp::socket* socket, boost::asio::streambuf* buf, const std::string& sep) {
+	boost::system::error_code error;
+	boost::asio::read_until(*socket, *buf, sep, error);
+	if (error && error != boost::asio::error::eof) {
+		printf("ProxyHandler: error! %s\n", error.message().c_str());
+	}
+	return error;
+}
+
+bool ProxyHandler::ReadNextHeader(std::istream* response_stream, std::string* header_key, std::string* header_value) {
+	std::string header;
+	if ( ! std::getline(*response_stream, header)) return false;
+	if (header == "\r") {
```

Comment:
`getline` will read until `\n`, which is guaranteed to terminate the header line. The `header == "\r"` check is checking that the header is _completely_ empty, which is indeed the end of the header section.